### PR TITLE
Fix using multiple decorators on the same instance

### LINKED ISSUE #77: Memoized methods of the same instance are mixed
If two functions are decorated of the same instance, their memoized values get mixed with the latest version (`v9.0.0`).

Take for example:

```typescript
class TestClass {
	@memDecorator()
	fn1() {
		return 1;
	}

	@memDecorator()
	counter2() {
		return 2;
	}
}

const instance = new TestClass();
console.log(instance.fn1()); // prints 1
console.log(instance.fn2()); // also prints 1 !!!
```

I would expect the second call to return 2, but it returns 1.

The memoized function in my understanding should be connected 1:1 to the _combination_ of _instance_ and _propterty_, whereas it is currently connected 1:1 to _instances_ (#74).

I guess a little bit more complicated internal data structure is needed instead of the current implementation of `decoratorInstanceMap`, will open a PR that adds a test and demonstrates a possible solution.

### PR DESCRIPTION
**Pull request by bencehornak**
Fixes #77

### CODE REVIEW

File: index.ts [hunk 1]
```diff
@@ -3,7 +3,18 @@ import mapAgeCleaner from 'map-age-cleaner';
...
+_propertyDescriptor_ to a memoized function.
+
+The memoized function that belongs to a specific instance and a specific
+propertyDescriptor could theoretically be accessed as follows:
+
+```
+const mem = decoratorInstanceMap.get(instance)?.get(propertyDescriptor)
+```
+ */
+const decoratorInstanceMap = new WeakMap<object, WeakMap<PropertyDescriptor, unknown>>(); // eslint-disable-line @typescript-eslint/ban-types
```
Comment (fregante) [inline index.ts:17]: I think the solution is to just move the weak map to the decorator function, but I'm not sure.

**Comment by sindresorhus** [comment]
// @Richienb

File: index.ts (hunk 1 above)
Comment (bencehornak) [reply index.ts]: That's actually a pretty good idea! Tested it, works!

File: index.ts [hunk 2]
```diff
@@ -169,6 +167,8 @@ export function memDecorator<
 >(
 	options: Options<FunctionToMemoize, CacheKeyType> = {},
 ) {
+	const decoratorInstanceMap = new WeakMap();
```
Comment (Richienb) [inline index.ts:170]:
```suggestion
	const instanceMap = new WeakMap();
```

File: test.ts [hunk 3]
```diff
@@ -239,19 +239,24 @@ test('prototype support', t => {
...
+			return returnValue2;
 		}
 	}
 
 	const alpha = new TestClass();
 	t.is(alpha.counter(), 1);
 	t.is(alpha.counter(), 1, 'The method should be memoized');
-
-	returnValue++;
+	t.is(alpha.counter2(), 101, 'Different methods should be independent');
```
Comment (Richienb) [inline test.ts:259]:
```suggestion
	t.is(alpha.counter2(), 101, 'The method should be memoized seperately from the other one');
```

**Review by Richienb — CHANGES_REQUESTED** [review]

**Pull request merged**